core/src/main/java/org/elasticsearch/action/fieldstats/FieldStatsRequest.java

```diff
@@ -33,17 +40,111 @@
+                            parseIndexContraints(indexConstraints, parser);
+                        } else {
+                            throw new IllegalArgumentException("Unknown field [" + fieldName + "]");
+                        }
+                        break;
+                    case START_ARRAY:
+                        if ("fields".equals(fieldName)) {
+                            while ((token = parser.nextToken()) != XContentParser.Token.END_ARRAY) {
+                                if (token.isValue()) {
+                                    fields.add(parser.text());
+                                }
+                            }
+                        } else {
+                            throw new IllegalArgumentException("Unknown field [" + fieldName + "]");
+                        }
+                        break;
+                    default:
+                        throw new IllegalArgumentException("Unknown token [" + token + "]");
+                }
+            }
+        }
+        this.fields = fields.toArray(new String[fields.size()]);
+        this.indexConstraints = indexConstraints.toArray(new IndexConstraint[indexConstraints.size()]);
+    }
+
+    private void parseIndexContraints(List<IndexConstraint> indexConstraints, XContentParser parser) throws IOException {
+        Token token = parser.currentToken();
+        assert token == Token.START_OBJECT;
+        String field = null;
+        String currentName = null;
+        for (token = parser.nextToken(); token != Token.END_OBJECT; token = parser.nextToken()) {
+            if (token == Token.FIELD_NAME) {
+                field = currentName = parser.currentName();
+            } else if (token == Token.START_OBJECT) {
+                for (token = parser.nextToken(); token != Token.END_OBJECT; token = parser.nextToken()) {
+                    if (token == Token.FIELD_NAME) {
+                        currentName = parser.currentName();
+                    } else if (token == Token.START_OBJECT) {
+                        IndexConstraint.Property property = IndexConstraint.Property.parse(currentName);
+                        token = parser.nextToken();
+                        if (token != Token.FIELD_NAME) {
+                            throw new IllegalArgumentException("Unexpected token [" + token + "]");
+                        }
+                        IndexConstraint.Comparison comparison = IndexConstraint.Comparison.parse(parser.currentName());
+                        token = parser.nextToken();
+                        if (token.isValue() == false) {
+                            throw new IllegalArgumentException("Unexpected token [" + token + "]");
+                        }
+                        String value = parser.text();
+                        indexConstraints.add(new IndexConstraint(field, property, comparison, value));
+
+                        token = parser.nextToken();
```

Comment:
Same here, this seems buggy (and any other place reusing the loop token var)? If the token is advanced here, then what happens when it is advanced in the loop iteration?